File: rental_store/backend/src/main/java/com/pata/keja/service/NotificationService.java
package com.pata.keja.service;

import com.pata.keja.enums.NotificationKind;
import com.pata.keja.dto.notification.NotificationResponse;
import com.pata.keja.dto.notification.NotificationSummaryResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;

public interface NotificationService {

    Page<NotificationResponse> listForUser(String userId, Pageable pageable);

    List<NotificationSummaryResponse> latestForUser(String userId);

    long unreadCount(String userId);

    NotificationResponse markRead(String notificationId, String userId);

    void markAllRead(String userId);

    /**
     * The single entry point used by every other service to emit a notification.
     */
    NotificationResponse emit(String userId,
            NotificationKind kind,
            String title,
            String body,
            String link);
}

File: rental_store/frontend/src/services/notificationService.ts
import { http } from "./apiClient";
import type { AppNotification } from "../types/notification";

interface Page<T> {
  content: T[];
}

interface NotificationResponse {
  id: string;
  kind: AppNotification["kind"];
  title: string;
  body: string;
  link: string | null;
  read: boolean;
  createdAt: string;
}

function toNotification(raw: NotificationResponse, userId: string): AppNotification {
  return {
    id: raw.id,
    userId,
    kind: raw.kind,
    title: raw.title,
    body: raw.body,
    read: raw.read,
    createdAt: raw.createdAt,
    ...(raw.link ? { link: raw.link } : {}),
  };
}

export const notificationService = {
  async listForUser(userId: string): Promise<AppNotification[]> {
    const page = await http.get<Page<NotificationResponse>>("/notifications");
    return page.content.map((notification) => toNotification(notification, userId));
  },

  async markRead(id: string): Promise<void> {
    await http.post<void>(`/notifications/${id}/read`);
  },

  async markAllRead(userId: string): Promise<void> {
    void userId;
    await http.post<void>("/notifications/read-all");
  },
};
